File: internal/runtime/tag.go
package runtime

import (
	"errors"
	"fmt"
	"io"
	"strings"
)

func cmdTag(args []string, cwd string, stdout, stderr io.Writer) int {
	name, err := parseTagArgs(args)
	if err != nil {
		fmt.Fprintln(stderr, err)
		return 2
	}
	workspace, ok := findWorkspace(cwd)
	if !ok {
		fmt.Fprintln(stderr, "not a feng workspace; run feng grow first")
		return 1
	}
	release, err := acquireWorkspaceLock(workspace, "tag")
	if err != nil {
		printJSON(stdout, map[string]any{"ok": false, "reason": "workspace_locked", "message": err.Error()})
		return 2
	}
	defer release()
	tag, err := createValidatedTag(workspace, name)
	if err != nil {
		fmt.Fprintf(stderr, "tag failed: %v\n", err)
		return 1
	}
	printJSON(stdout, map[string]any{"ok": true, "tag": tag, "commit": currentHead(workspace)})
	return 0
}

func parseTagArgs(args []string) (string, error) {
	if len(args) != 1 || strings.TrimSpace(args[0]) == "" {
		return "", errors.New("tag requires exactly one name")
	}
	name := strings.TrimSpace(args[0])
	if slug(name) != name {
		return "", errors.New("tag name must contain only letters, numbers, dot, dash, or underscore")
	}
	return name, nil
}

func createValidatedTag(workspace, name string) (string, error) {
	state, err := loadState(workspace)
	if err != nil {
		return "", err
	}
	if state.CandidateStatus != "validated" || state.ValidatedCommit == "" {
		return "", errors.New("tag requires candidate_status=validated; run feng check first")
	}
	if currentHead(workspace) != state.ValidatedCommit {
		return "", errors.New("tag requires HEAD to match the validated commit; run feng check first")
	}
	status, err := selfGitStatus(workspace)
	if err != nil {
		return "", err
	}
	if strings.TrimSpace(status) != "" {
		return "", errors.New("tag requires clean feng self roots")
	}
	if _, err := runGit(workspace, "rev-parse", "--verify", "refs/tags/"+name); err == nil {
		return "", errors.New("tag already exists: " + name)
	}
	if _, err := runGit(workspace, "tag", name, state.ValidatedCommit); err != nil {
		return "", err
	}
	appendEvent(workspace, "tag_created", map[string]any{"tag": name, "commit": state.ValidatedCommit})
	return name, nil
}

func currentValidatedTag(workspace, commit string) string {
	if commit == "" {
		return ""
	}
	out, err := runGit(workspace, "tag", "--points-at", commit)
	if err != nil {
		return ""
	}
	for _, line := range strings.Split(out, "\n") {
		tag := strings.TrimSpace(line)
		if tag != "" {
			return tag
		}
	}
	return ""
}

File: src/feng/tag.py
from __future__ import annotations

import re
from pathlib import Path

from .events import append_event
from .git_utils import current_head, git, self_status_short
from .state import load_state


TAG_RE = re.compile(r"[A-Za-z0-9._-]+")


def create_validated_tag(workspace: Path, name: str) -> str:
    name = name.strip()
    if not TAG_RE.fullmatch(name):
        raise RuntimeError("tag name must contain only letters, numbers, dot, dash, or underscore")
    state = load_state(workspace)
    validated_commit = state.get("validated_commit", "")
    if state.get("candidate_status") != "validated" or not validated_commit:
        raise RuntimeError("tag requires candidate_status=validated; run feng check first")
    if current_head(workspace) != validated_commit:
        raise RuntimeError("tag requires HEAD to match the validated commit; run feng check first")
    if self_status_short(workspace):
        raise RuntimeError("tag requires clean feng self roots")
    code, _ = git(workspace, "rev-parse", "--verify", f"refs/tags/{name}", timeout=10)
    if code == 0:
        raise RuntimeError(f"tag already exists: {name}")
    code, output = git(workspace, "tag", name, validated_commit, timeout=30)
    if code != 0:
        raise RuntimeError(output.strip() or "git tag failed")
    append_event(workspace, "tag_created", {"tag": name, "commit": validated_commit})
    return name
